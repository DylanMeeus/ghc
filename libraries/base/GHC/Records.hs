{-# LANGUAGE NoImplicitPrelude
           , MultiParamTypeClasses
           , MagicHash
           , KindSignatures
           , DataKinds
           , FunctionalDependencies
  #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  GHC.Records
-- Copyright   :  (c) Adam Gundry 2015
-- License     :  see libraries/base/LICENSE
--
-- Maintainer  :  cvs-ghc@haskell.org
-- Stability   :  internal
-- Portability :  non-portable (GHC extensions)
--
--
-----------------------------------------------------------------------------

module GHC.Records
       ( HasField(..)
       ) where

import GHC.Base ( Symbol )
import GHC.Exts ( Proxy# )

class HasField (x :: Symbol) r a | x r -> a where
  getField :: Proxy# x -> r -> a
