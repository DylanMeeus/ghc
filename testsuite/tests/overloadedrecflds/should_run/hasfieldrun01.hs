{-# LANGUAGE DataKinds, FlexibleInstances, MagicHash,
             MultiParamTypeClasses, TypeFamilies #-}

import GHC.Prim (Proxy#, proxy#)
import GHC.Records (HasField(..))

type family B where B = Bool

data T = MkT { foo :: Int, bar :: B }

data U a b = MkU { baf :: a }

t = MkT 42 True

u :: U Char Char
u = MkU 'x'

-- A virtual foo field for U
instance HasField "foo" (U a b) [Char] where
  getField _ _ = "virtual"

main = do print (getField (proxy# :: Proxy# "foo") t)
          print (getField (proxy# :: Proxy# "bar") t)
          print (getField (proxy# :: Proxy# "baf") u)
          print (getField (proxy# :: Proxy# "foo") u)
