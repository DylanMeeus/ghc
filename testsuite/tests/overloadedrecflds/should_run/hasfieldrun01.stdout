42
True
'x'
"virtual"
